File: programs/cdp_stablecoin_protocol/src/instructions/update_interest_rate.rs
use crate::{
    constants::{BPS_SCALE, INTEREST_SCALE, MAX_INTEREST_RATE, MIN_INTEREST_RATE, YEAR_IN_SECONDS},
    errors::ArithmeticError,
    state::ProtocolConfig,
};
use anchor_lang::prelude::*;
use anchor_spl::{
    associated_token::AssociatedToken,
    token::{Mint, Token, TokenAccount},
};
use pyth_solana_receiver_sdk::price_update::{get_feed_id_from_hex, PriceUpdateV2};

#[derive(Accounts)]
pub struct UpdateInterestRate<'info> {
    #[account(mut)]
    user: Signer<'info>,
    collateral_mint: Account<'info, Mint>,

    #[account(
        mut,
        seeds = [b"config"],
        bump = protocol_config.bump
    )]
    pub protocol_config: Account<'info, ProtocolConfig>,

    #[account(owner = pyth_solana_receiver_sdk::ID)]
    pub stablecoin_price_feed: Account<'info, PriceUpdateV2>,
    #[account(
        mut,
        address = protocol_config.stable_mint,
        mint::decimals = 6,
        mint::authority = auth,
    )]
    stable_mint: Account<'info, Mint>,
    #[account(
        mut,
        seeds = [b"treasury", stable_mint.key().as_ref()],
        token::mint = stable_mint,
        token::authority = auth,
        bump
    )]
    pub treasury_vault: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        seeds = [b"stake_vault", stable_mint.key().as_ref(), collateral_mint.key().as_ref()],
        token::mint = stable_mint,
        token::authority = auth,
        bump
    )]
    pub stake_vault: Box<Account<'info, TokenAccount>>,
    /// CHECK: This is an auth acc for the vault
    #[account(
        mut,
        seeds = [b"auth"],
        bump = protocol_config.auth_bump
    )]
    auth: UncheckedAccount<'info>,
    token_program: Program<'info, Token>,
    associated_token_program: Program<'info, AssociatedToken>,
    system_program: Program<'info, System>,
}

impl<'info> UpdateInterestRate<'info> {
    // Taylor series approximation of e^x up to 3rd order
    fn exponential_approximation(x: i128, scale: u128) -> Result<u128> {
        // first term: 1
        let mut result = scale as i128;

        // second term: x
        result = result
            .checked_add(x)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        // third term: x^2/2!
        let x_squared = x
            .checked_mul(x)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        let x_squared_scaled = x_squared
            .checked_div(scale as i128)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        result = result
            .checked_add(x_squared_scaled / 2)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        // fourth term: x^3/3!
        let x_cubed = x_squared_scaled
            .checked_mul(x)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        let x_cubed_scaled: i128 = x_cubed
            .checked_div(scale as i128)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        let x_cubed_final = x_cubed_scaled / 6;

        result = result
            .checked_add(x_cubed_final)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        Ok(result as u128)
    }

    // Calculate interest rate based on price deviation from peg
    fn calculate_interest_rate(
        stablecoin_price: i64,
        stablecoin_exponent: i32,
        base_rate_bps: u16,
        sigma_bps: u16,
    ) -> Result<u128> {
        // Calculate price deviation from peg
        let peg = 10_i128.pow(stablecoin_exponent.abs() as u32);

        let price_deviation: i128 = peg
            .checked_sub(stablecoin_price as i128)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        // exponent calculation
        let bps_scale_i128 = BPS_SCALE as i128;

        let mut x = price_deviation
            .checked_mul(bps_scale_i128)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        x = x / sigma_bps as i128;

        // convert exponent from price scale to interest scale
        x = x
            .checked_mul(INTEREST_SCALE as i128)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        x = x / peg;

        // Calculate rate = base_rate * e^x
        let exp_result = Self::exponential_approximation(x, INTEREST_SCALE)?;

        let interest_rate: u128 = exp_result
            .checked_mul(base_rate_bps as u128)
            .ok_or(ArithmeticError::ArithmeticOverflow)?
            .checked_div(BPS_SCALE as u128)
            .ok_or(ArithmeticError::ArithmeticOverflow)?;

        Ok(interest_rate.clamp(MIN_INTEREST_RATE, MAX_INTEREST_RATE))
    }

    // Calculate compound interest using Taylor series for efficiency for (1+r)^t
    fn compound_interest(interest_rate: u128, time_elapsed: u128) -> Result<u128> {
        // Handle small periods (t <= 4) directly for better precision
        match time_elapsed {
            0 => return Ok(INTEREST_SCALE),
            1 => return Ok(INTEREST_SCALE + interest_rate),
            2 => {
                return Ok(
                    (INTEREST_SCALE + interest_rate) * (INTEREST_SCALE + interest_rate)
                        / INTEREST_SCALE,
                )
            }
            3 => {
                return Ok((INTEREST_SCALE + interest_rate)
                    * (INTEREST_SCALE + interest_rate)
                    * (INTEREST_SCALE + interest_rate)
                    / (INTEREST_SCALE * INTEREST_SCALE))
            }
            4 => {
                let pow_two = (INTEREST_SCALE + interest_rate) * (INTEREST_SCALE + interest_rate)
                    / INTEREST_SCALE;
                return Ok((pow_two * pow_two) / INTEREST_SCALE);
            }
            _ => (),
        }

        // For larger periods use Taylor expansion
        let exp = time_elapsed;
        let exp_minus_one = time_elapsed - 1;
        let exp_minus_two = time_elapsed - 2;

        let base = interest_rate;
        let base_pow_two = interest_rate * interest_rate / INTEREST_SCALE;
        let base_pow_three = base_pow_two * base / INTEREST_SCALE;

        let first_term = INTEREST_SCALE; // denoting 1
        let second_term = exp * base;
        let third_term = exp * exp_minus_one * base_pow_two / 2;
        let fourth_term = exp * exp_minus_one * exp_minus_two * base_pow_three / 6;

        Ok(first_term + second_term + third_term + fourth_term)
    }

    // Main function to update protocol interest rate
    pub fn update_interest_rate(&mut self) -> Result<()> {
        // Calculate time elapsed
        let current_timestamp = Clock::get()?.unix_timestamp;
        let time_elapsed =
            (current_timestamp - self.protocol_config.last_interest_rate_update_time) as u64;

        match time_elapsed {
            0 => return Ok(()),
            _ => {}
        };

        // Get current stablecoin price
        let price_feed = &self.stablecoin_price_feed;

        let feed_id: [u8; 32] = get_feed_id_from_hex(&self.protocol_config.stablecoin_price_feed)?;

        let stablecoin_price = price_feed.get_price_unchecked(&feed_id)?;

        // Calculate yearly interest rate

        let new_interest_rate_yearly = Self::calculate_interest_rate(
            stablecoin_price.price,
            stablecoin_price.exponent,
            self.protocol_config.base_rate,
            self.protocol_config.sigma,
        )?;
        self.protocol_config.last_interest_rate = new_interest_rate_yearly;

        // Calculate compound interest
        let per_second_rate = new_interest_rate_yearly / YEAR_IN_SECONDS as u128;

        let compounded_interest_rate =
            Self::compound_interest(per_second_rate, time_elapsed as u128)?;
        
        // Calculate interest revenue
        let old_total_debt = self.protocol_config.total_debt as u128;
        let new_total_debt = old_total_debt
            .checked_mul(compounded_interest_rate)
            .ok_or(ArithmeticError::ArithmeticOverflow)?
            / INTEREST_SCALE;
        let interest_revenue = new_total_debt
            .checked_sub(old_total_debt)
            .ok_or(ArithmeticError::ArithmeticOverflow)? as u64;
        
        // Distribute revenue
        self.protocol_config.distribute_revenue(
            interest_revenue,
            &self.stable_mint,
            &self.treasury_vault,
            &self.stake_vault,
            &self.auth,
            &self.token_program,
        )?;

        // Update protocol state
        self.protocol_config.cumulative_interest_rate =
            (self.protocol_config.cumulative_interest_rate)
                .checked_mul(compounded_interest_rate)
                .ok_or(ArithmeticError::ArithmeticOverflow)?
                / INTEREST_SCALE;

        self.protocol_config.total_debt = new_total_debt;
        self.protocol_config.last_interest_rate_update_time  = current_timestamp;
        
        Ok(())
    }
}

File: research/protocol_model/src/instructions/update_interest_rate.py
"""Interest rate update logic - direct port from Rust"""
import time
from typing import Optional, Tuple
from ..state.protocol_config import ProtocolConfig
from ..errors import ArithmeticError
from ..constants import (
    MIN_INTEREST_RATE,
    MAX_INTEREST_RATE,
    PRICE_SCALE,
    INTEREST_SCALE,
    BPS_SCALE,
    YEAR_IN_SECONDS
)

def exponential_approximation(x: int, scale: int) -> int:
    """Port of Rust exponential_approximation
    Using Taylor series of order 3
    All values scaled by INTEREST_SCALE
    """
    # first term: 1
    result = scale

    # second term: x
    result = checked_add(result, x)

    # third term: x^2/2!
    x_squared = checked_mul(x, x) // scale

    result = checked_add(result, x_squared // 2)

    # fourth term: x^3/3!
    x_cubed = checked_mul(x_squared, x) // scale

    result = checked_add(result, x_cubed // 6)

    return result

def calculate_interest_rate(
    stablecoin_price: int,  # i64 in Rust
    base_rate_bps: int,     # u16 in Rust
    sigma_bps: int          # u16 in Rust
) -> int:                   # Returns u128 in Rust
    """Calculate interest rate using exponential approximation"""
    # Calculate price deviation from peg
    price_deviation = PRICE_SCALE - stablecoin_price
    if price_deviation < 0:
        raise ArithmeticError("Price deviation underflow")

    # Calculate exponent
    # First in price scale
    x = checked_mul(price_deviation, BPS_SCALE) // sigma_bps

    # Convert exponent from price scale to interest scale
    x = checked_mul(x, INTEREST_SCALE) // PRICE_SCALE

    # Calculate e^x and multiply by base rate
    exp_term = exponential_approximation(x, INTEREST_SCALE)
    interest_rate = checked_mul(exp_term, base_rate_bps) // BPS_SCALE

    # Clamp to min/max
    return min(max(interest_rate, MIN_INTEREST_RATE), MAX_INTEREST_RATE)

def compound_interest(interest_rate: int, time_elapsed: int) -> int:
    """Port of Rust compound_interest
    All values scaled by INTEREST_SCALE
    """
    # Handle small time periods directly
    if time_elapsed == 0:
        return INTEREST_SCALE
    elif time_elapsed == 1:
        return INTEREST_SCALE + interest_rate
    elif time_elapsed == 2:
        one_plus_rate = INTEREST_SCALE + interest_rate
        return checked_mul(one_plus_rate, one_plus_rate) // INTEREST_SCALE
    elif time_elapsed == 3:
        one_plus_rate = INTEREST_SCALE + interest_rate
        squared = checked_mul(one_plus_rate, one_plus_rate) // INTEREST_SCALE
        return checked_mul(squared, one_plus_rate) // INTEREST_SCALE
    elif time_elapsed == 4:
        one_plus_rate = INTEREST_SCALE + interest_rate
        squared = checked_mul(one_plus_rate, one_plus_rate) // INTEREST_SCALE
        return checked_mul(squared, squared) // INTEREST_SCALE

    # For larger time periods use Taylor expansion
    exp = time_elapsed
    exp_minus_one = time_elapsed - 1
    exp_minus_two = time_elapsed - 2

    base = interest_rate
    base_pow_two = checked_mul(base, base) // INTEREST_SCALE
    base_pow_three = checked_mul(base_pow_two, base) // INTEREST_SCALE

    first_term = INTEREST_SCALE
    second_term = checked_mul(exp, base)
    third_term = checked_mul(checked_mul(exp, exp_minus_one), base_pow_two) // 2
    fourth_term = checked_mul(
        checked_mul(
            checked_mul(exp, exp_minus_one),
            exp_minus_two
        ),
        base_pow_three
    ) // 6

    return checked_add(
        first_term,
        checked_add(
            second_term,
            checked_add(third_term, fourth_term)
        )
    )

def checked_mul(a: int, b: int) -> int:
    """Multiply with overflow checking"""
    result = a * b
    if result > 2**128 - 1:  # u128::MAX
        raise ArithmeticError("Arithmetic overflow in multiplication")
    return result

def checked_div(a: int, b: int) -> int:
    """Divide with overflow checking"""
    if b == 0:
        raise ArithmeticError("Division by zero")
    return a // b

def checked_add(a: int, b: int) -> int:
    """Add with overflow checking"""
    result = a + b
    if result > 2**128 - 1:  # u128::MAX
        raise ArithmeticError("Arithmetic overflow in addition")
    return result

def get_mock_price() -> Tuple[int, int]:
    """Mock function to get price from Pyth
    Returns:
        Tuple[price, exponent] - price with exponent -6 (6 decimal places)
    """
    return (1_000_000, -6)  # $1.00 with 6 decimals

def update_interest_rate(
    protocol_config: ProtocolConfig,
    mock_price: Optional[Tuple[int, int]] = None,
    time_elapsed: Optional[int] = None
) -> None:
    """Port of Rust update_interest_rate"""
    # Calculate time elapsed
    if time_elapsed is None:
        current_timestamp = int(time.time())
        time_elapsed = current_timestamp - protocol_config.last_interest_rate_update
    if time_elapsed == 0:
        return

    # Get current stablecoin price
    price, _ = mock_price if mock_price is not None else get_mock_price()
    
    # Calculate yearly interest rate
    new_interest_rate_yearly = calculate_interest_rate(
        price,
        protocol_config.base_rate,
        protocol_config.sigma
    )

    # Convert to per-second rate and compound
    new_interest_rate = new_interest_rate_yearly // YEAR_IN_SECONDS
    compounded_interest_rate = compound_interest(new_interest_rate, time_elapsed)

    # Update protocol state
    protocol_config.cumulative_interest_rate = checked_mul(
        protocol_config.cumulative_interest_rate,
        compounded_interest_rate
    ) // INTEREST_SCALE
    
    protocol_config.total_debt = checked_mul(
        protocol_config.total_debt,
        compounded_interest_rate
    ) // INTEREST_SCALE
    
    protocol_config.last_interest_rate_update = (
        int(time.time()) if time_elapsed is None 
        else protocol_config.last_interest_rate_update + time_elapsed
    ) 